sound/soc/sof/intel/hda-ipc.c

```diff
@@ -142,19 +157,12 @@ irqreturn_t hda_dsp_ipc_irq_thread(int irq, void *context)
 
 		/* handle immediate reply from DSP core - ignore ROM messages */
 		if (msg != 0x1004000)
-			snd_sof_ipc_reply(sdev, msg);
+			reply = snd_sof_ipc_reply(sdev, msg);
 
-		/* clear DONE bit - tell DSP we have completed the operation */
-		snd_sof_dsp_update_bits_forced(sdev, HDA_DSP_BAR,
-					       HDA_DSP_REG_HIPCIE,
-					       HDA_DSP_REG_HIPCIE_DONE,
-					       HDA_DSP_REG_HIPCIE_DONE);
-
-		/* unmask Done interrupt */
-		snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR,
-					HDA_DSP_REG_HIPCCTL,
-					HDA_DSP_REG_HIPCCTL_DONE,
-					HDA_DSP_REG_HIPCCTL_DONE);
+		/* handle immediate reply from DSP core. If the msg is */
+		/* found, set done bit in reply_done, else set it here */
+		if (reply)
+			hda_dsp_ipc_cmd_done(sdev, SOF_IPC_DSP_REPLY);
```

Comment:
is this what you want to do when the reply is -EINVAL?

Reply:
yes, if the return value is 0,  the cmd_done would be called at the end of msg processing function.
But if is  -EINVAL, no msg processing function would be called, so it could be called at here.